### PR TITLE
Adapt api to cosmos

### PR DESCRIPTION
**Pull request by rklaehn**
The biggest change is that an empty tree does not need a secret, so we can have a default instance.

Also, some module restructuring to hopefully make things simpler.

image

### CODE REVIEW

File: banyan/Cargo.toml [hunk 1]
```diff
@@ -21,7 +21,7 @@ parking_lot = "0.11.1"
 smallvec = "1.6.1"
 tracing = "0.1.25"
 weight-cache = "0.2.3"
-zstd = "0.8.0"
```
Comment (rklaehn) [inline banyan/Cargo.toml:24]: Had to revert the update - since I had some weird perf problems with 0.8.0, and I don't want to look into those right now.

**Review by dvc94ch — APPROVED** [review]
lgtm

File: banyan/src/stream_builder.rs [hunk 2]
```diff
@@ -158,6 +170,21 @@ impl<T: TreeTypes> StreamBuilder<T> {
         self.root.as_ref().map(|x| x.as_ref())
     }
 
+    /// Modify a StreamBuilder and roll back the changes if the operation was not successful
+    ///
+    /// Note that consumed offets are *not* rolled back to make sure we don't reuse offsets.
```
Comment (wngr) [inline banyan/src/stream_builder.rs:175]: Does that mean that there might be offset gaps now within the tree? So a chunk is 100..200 but only `data.len()` is 50?

**Review by wngr — APPROVED** [review]

File: banyan/src/stream_builder.rs (hunk 2 above)
Comment (rkuhn) [reply banyan/src/stream_builder.rs:175]: yeah, I was similarly confused: these are different offsets (namely bytes written with the stream cipher) — these new offsets should definitely be named differently

Comment (wngr) [reply banyan/src/stream_builder.rs:175]: I see, that makes sense.

:+1: for renaming. I saw `StreamOffset` somewhere? Maybe `CipherOffset`/`CryptoOffset`/`EncryptionOffset`? `CipherPointer`?

Comment (rklaehn) [reply banyan/src/stream_builder.rs:175]: Yes, you can have gaps. When you build a tree you won't have any gaps, but as soon as you do things like pack or retain, there will be gaps. Nothing wrong with that though, you are not going to run out of u64 any time soon, and gaps are not a problem.

`CIpherOffset` sounds reasonable. Will rename.

Comment (rkuhn) [reply banyan/src/stream_builder.rs:175]: … and if someone is about to run out of u64 cipher offsets, we can rekey the tree. But we need to hurry, at 100MB/s we only have about 5700 years before this happens 😀.

Comment (rklaehn) [reply banyan/src/stream_builder.rs:178]: I was going to add methods for rekeying in any case, but it is pretty straightforward and not that urgent...

**Pull request merged**